### PR TITLE
Add streaming version of XxHash64

### PR DESCRIPTION
**Pull request by electrum**

### CODE REVIEW

File: src/main/java/io/airlift/slice/XxHash64.java [hunk 1]
```diff
@@ -69,8 +69,8 @@ public static long hash(long seed, Slice data, int offset, int length)
         hash += length;
 
         // round to the closest 32 byte boundary
-        // this is the point up to which {@see #updateBody} processed
-        int index = length & 0xFFFFFFE0;
```
Comment (martint) [inline src/main/java/io/airlift/slice/XxHash64.java]: Don't do this. This is a mask, and `-32` makes it hard to tell what it is.

File: src/main/java/io/airlift/slice/XxHash64.java [hunk 2]
```diff
@@ -56,7 +56,7 @@ public static long hash(long seed, Slice data, int offset, int length)
         checkPositionIndexes(0, offset + length, data.length());
 
         Object base = data.getBase();
-        final long address = data.getAddress() + offset;
```
Comment (martint) [inline src/main/java/io/airlift/slice/XxHash64.java]: This was intentional, to make sure the address variable is not inadvertently overwritten as the code walks loops through the data.

**Comment by martint** [comment]
Have you run the benchmarks to make sure your changes didn't affect performance for the non-streaming version? The structure of the code was very carefully tuned for play nice with how hotspot inlines/optimizes it.

**Comment by martint** [comment]
I'm not sure I like mixing the static vs non-static methods in the same class. I'd probably introduce a separate class for the streaming version

File: src/main/java/io/airlift/slice/XxHash64.java (hunk 1 above)
Comment (electrum) [inline src/main/java/io/airlift/slice/XxHash64.java]: I think is the standard way to write it and is more clear (the JDK does it this way), but I can leave it.

**Comment by martint** [comment]
looks good

**Comment by electrum** [comment]
I found that computing the body from `v1`-`v4` in a method caused a substantial performance penalty, so I removed that. Extracting a method for `updateTail` was the same performance. I moved that into its own commit to make it easy to benchmark in isolation.

**Pull request merged**